runner/llamarunner/cache.go

```diff
@@ -231,16 +231,38 @@ func (c *InputCache) ShiftCacheSlot(slot *InputCacheSlot, numKeep int) error {
 	slog.Debug("context limit hit - shifting", "id", slot.Id, "limit", c.numCtx, "input", len(slot.Inputs),
 		"keep", numKeep, "discard", discard)
 
-	// TODO (jessegross): KV cache removal can fail for certain types of models
-	if !c.lc.KvCacheSeqRm(slot.Id, numKeep, numKeep+discard) {
-		return fmt.Errorf("unable to remove old kv cache entries (id: %v, keep: %v discard: %v)", slot.Id, numKeep, discard)
-	}
-	c.lc.KvCacheSeqAdd(slot.Id, numKeep+discard, len(slot.Inputs), -discard)
+	if c.lc.KvCacheCanShift() {
+		if !c.lc.KvCacheSeqRm(slot.Id, numKeep, numKeep+discard) {
+			return fmt.Errorf("unable to remove old kv cache entries (id: %v, keep: %v discard: %v)", slot.Id, numKeep, discard)
+		}
+		c.lc.KvCacheSeqAdd(slot.Id, numKeep+discard, len(slot.Inputs), -discard)
+
+		for i := numKeep + discard; i < len(slot.Inputs); i++ {
+			slot.Inputs[i-discard] = slot.Inputs[i]
+		}
+		slot.Inputs = slot.Inputs[:len(slot.Inputs)-discard]
+	} else {
+		slog.Debug("kv cache cannot shift, clearing cache and truncating history")
+
+		// Clear the entire KV cache
+		if !c.lc.KvCacheSeqRm(slot.Id, 0, -1) {
+			return fmt.Errorf("unable to remove kv cache entries (id: %v)", slot.Id)
+		}
+
+		// Update the slot.Inputs to match what would happen with a shift operation
+		// Keep the first numKeep tokens and the tokens after the discard
+		keepInputs := make([]input, numKeep)
+		copy(keepInputs, slot.Inputs[:numKeep])
+
+		afterDiscardInputs := make([]input, len(slot.Inputs)-(numKeep+discard))
+		copy(afterDiscardInputs, slot.Inputs[numKeep+discard:])
 
-	for i := numKeep + discard; i < len(slot.Inputs); i++ {
-		slot.Inputs[i-discard] = slot.Inputs[i]
+		// Update the inputs to match what would happen after a shift
+		newInputs := make([]input, 0, numKeep+len(afterDiscardInputs))
+		newInputs = append(newInputs, keepInputs...)
+		newInputs = append(newInputs, afterDiscardInputs...)
+		slot.Inputs = newInputs
```

Comment:
I think this is actually updating the wrong set of inputs. `slot.Inputs` reflects what is stored in the cache, however, since we just cleared it, `newInputs` are not actually in there.

`Sequence.inputs` is our pending input queue to process so if we prepend `newInputs` to that then we can regenerate the cache state so that it looks the same as if we shifted - this will require a pass through the transformer. When that happens `slot.Inputs` will be updated to reflect the correct state through its normal process.